### PR TITLE
[4.0] webcomponents everywhere, admin menu, dropdown

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

## Glimpse of Joomla 4 fully componentized 

This PR moves the menu script from the template to the respected module (mod_menu) and adds the dropdown script for the status module (mod_status, on the top right corner).
Both scripts are custom elements!!!

So now the dashboard is the first page in the administrator area the implements fully custom elements!

#### What does it mean?
Well there is a reduction in the scripts size over **250%** which of course results to an exactly same amount of reduction in the parse time and execution time of the browser!!! **Do the math...**

And here's the proof:
**Joomla 3.7.5, total scripts size: 148.791Kb**
screen shot 2017-09-04 at 20 17 35

**Joomla 4.0 pre-alpha, total scripts size: 57.24Kb**
screen shot 2017-09-04 at 20 16 46

### Testing Instructions

Apply patch, menu and top right links (user & bell icon) still functioning as before

### CODE REVIEW

**Comment by C-Lodder** [comment]
Update checks arent even minified, so scripts size is even smaller :)

**Comment by alikon** [comment]
i like that (**250**%) number
i hope to see "shortly" same number from frontend too
 :punch:

**Comment by dgrammatiko** [comment]
@alikon expect more than 300% reduction for the front end scripts size. Joomla 4 will be a good base for delivering mobile ready sites, as long as 3rd PD follow the same path

File: administrator/modules/mod_menu/tmpl/default.php [hunk 1]
```diff
@@ -9,6 +9,7 @@
 
 defined('_JEXEC') or die;
 
+$siteName  = \Joomla\CMS\Factory::getApplication()->get('sitename');
```
Comment (brianteeman) [inline administrator/modules/mod_menu/tmpl/default.php:12]: you need to escape and sanitise the sitename see #17209 for reason and example code

File: administrator/templates/atum/index.php [hunk 2]
```diff
@@ -127,28 +113,34 @@
 			<?php endif; ?>
 			<section id="content" class="content">
 				<?php // Begin Content ?>
-				<jdoc:include type="modules" name="top" style="xhtml" />
+				<jdoc:include type="modules" name="top" style="xhtml"/>
```
Comment (brianteeman) [inline administrator/templates/atum/index.php:116]: xhtml ?

Comment (C-Lodder) [reply administrator/templates/atum/index.php:116]: We'll be sorting that out in another PR

Comment (brianteeman) [reply administrator/templates/atum/index.php:116]: ok

**Comment by C-Lodder** [comment]
Nearly there. Basic functionality is working fine, however navigating via a keyboard seems a bit broken.

Compare against the 4.0-dev branch and you'll see what I mean

**Comment by dgrammatiko** [comment]
@C-Lodder most probably is coming from wrong context (the keyboard listening functions is bound to the wrong element). Should be easy to fix.
There is is bigger issue that needs to be solved here: the workflow for web components/custom elements that are not part of the other repo, do we:
- create a srcWebComponents folder (or something similar) in the root
- deal in the other repo and just import
- beef up grunt to deal with the source (same folder, reads a specific file and creates the minified es6 + the transpiled ones)

@wilsonge @mbabker any input will be welcome here

**Comment by mbabker** [comment]
If you need to put raw undistributed resources into this repo use the `build` folder, that already has a plethora of stuff.

**Comment by C-Lodder** [comment]
@dgt41 - I'd say lets put them in the other repo but in `src/joomla`. We can then continue to work on them as normal. Then beef up grunt a little and use the following paths when pulling them in:

**official components**
> media/vendor/webcomponents/xxx

**components for Joomla**
> media/system/js/webcomponents/xxx

thoughts?

**Comment by dgrammatiko** [comment]
@C-Lodder importing **all** scripts has a huge negative in the PR workflow, someone needs to clone Joomla repo **and** the custom element repo. Do the work in the ce repo and copy paste to the actual Joomla folder to test. The basic UI components its ok to be another repo, but fields and components/modules/plugins need to stay in the same repo for easier bug fixing/code improvements. TBH I'm doing that for some time and it's way too hard.

> media/system/js/webcomponents/xxx

That also cannot be good option for few reasons:
- web components use the relative path code in JHtml, means that the correct path would be `media/system/webcomponents/xxx`, Note there is no `js` folder involved
- adding uncompiled scripts to the above location, even if we beef up grunt etc, we are going to deliver uncompiled scripts as well, not a good idea (or we need to unnecessarily complicate the build script...)
- we need to differentiate old ES5 script from ES6 as much as possible, think ESLint, Hound, airBnB CS

I think `build/webcomponents` is a good option in conjunction with the beefed up grunt (or shall we go to gulp or even web pack? Although such a move is a pretty big task on it's own)

**Review by yvesh — APPROVED** [review]

**Comment by roland-d** [comment]
I have tested this item :white_check_mark: successfully on a14863ffb9ab7aa0ef05b1210586e6bb9e2ea8cc<br /><br />Here are the sizes before and after applying the patch

Page          | Before Chrome / FireFox | After Chrome / FireFox
Control Panel | 219KB / 229.15KB        | 62.3KB / 72.98KB
Article list  | 200KB / 209.36KB        | 200KB / 208.81KB
Category list | 241KB / 249.72KB        | 241KB / 249.17KB
User list     | 231KB / 239.96KB        | 231KB / 239.41KB
Extension Man | 200KB / 209.36KB        | 200KB / 208.81KB

So there is a huge gain on the control panel, other pages it has no effect but I think this is as expected but it would be great if the same can be done for other pages.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17868">issues.joomla.org/tracker/joomla-cms/17868</a>.</sub>

**Comment by fancyFranci** [comment]
I have tested this item :red_circle: unsuccessfully on a14863ffb9ab7aa0ef05b1210586e6bb9e2ea8cc<br /><br />Scripts are a lot smaller, yes! But the bell and user icon on the top right corner are not working anymore. Before the patch they showed their boxes. Now when I click on them, nothing happens :/
Windows Chrome v62.0.3202.62 and Firefox v56.0.1, testet on Joomla 4.0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17868">issues.joomla.org/tracker/joomla-cms/17868</a>.</sub>

**Pull request closed without merging**